### ISSUE TITLE
"Max-Tokens" API setting isn't getting included in the chat request body

### ISSUE THREAD

**Issue by cheeseonamonkey** [jackschedel/gpt.koaladev.io#34]
Just notice that the *"**Max-Tokens**"* API setting isn't getting included in the chat request body, even though it appears to be a user configuration setting, and is likely being stored.

![image](https://github.com/jackschedel/gpt.koaladev.io/assets/54555500/9897fdde-b944-4f60-a6ab-7f31f1df3fc2)

---

- *Max Tokens parameter is a control for the maximum number of tokens that can be generated in a single call to the GPT model*

**Comment by jackschedel** [comment]
its worth noting that simple bugs like these would probably also be accepted on the original repo aswell. I've made a few PRs to the original repo.

**Comment by cheeseonamonkey** [comment]
It doesn't appear to be well-maintained, but yes you are right :)

**Comment by ghost** [comment]
To expand on that, "Max Token" *is* actually used by the frontend, but not in the way it is described - it actually controls the maximum possible context for requests (you can see that [here](https://github.com/jackschedel/gpt.koaladev.io/blob/main/src/hooks/useSubmit.ts#L73)), so if there are older messages that do not fit in the context, they'll be cut off, you can test that by having a big enough conversation and changing "Max Token" to like 500.

Ideally there should be two separate settings - one for controlling the context size, the other for `max_tokens` parameter to pass to OpenAI.

**Comment by jackschedel** [comment]
good to know. I'll make sure to add them as separate options rather than just changing the behavior of it

**Comment by ghost** [comment]
Thanks! Changing the max context size is mainly useful where you don't care that much about AI remembering a lot of your chat's history, and want cheaper requests, while `max_tokens` is kind of more useful since it strongly affects the length of the AI response (without having to say it in the prompt). If `max_tokens` is not provided, the API will generate as many tokens as available for the current context (basically `model_context_size - passed_messages_token_count`). With `max_tokens` it'll have to be included in the context calculation, although that's trivial to add - just change `chats[currentChatIndex].config.max_tokens` in the call to `limitMessageTokens` to something like `chats[currentChatIndex].config.max_context - chats[currentChatIndex].config.max_tokens` (assuming the options get changed).

**Comment by cheeseonamonkey** [comment]
@Yardanico 
~~The request isn't sent in this application because the client-side understands that it is too many tokens, and doesn't send it *(providing an error message instead)*.~~
But this is the result of a raw CURL:
![image](https://github.com/jackschedel/gpt.koaladev.io/assets/54555500/1bd420e9-063d-4939-9c76-1abc166ec2d7)
~~It seems to return a rejected HTTP error response.~~
~~Additionally, I do not see any parameters except for max_tokens in the [documentation](https://platform.openai.com/docs/api-reference/chat/create).~~

Entirely possible also that I'm misunderstanding something 😅

---

I _was_ misunderstanding. This seems like a great feature.

You are saying that the maximum token count in the request will be calculated using 2 client settings:
- context size / max context to include *(how much chat history to include)*
- max desired response *(assistant's response size)*

And:
- the size of the most recent message 

As is currently, I occasionally find myself doing this in my head anyway. And the only purpose for setting `max_token` is perhaps to avoid infinite looping. 

These would make an additional setting, but result in settings being actually useful, and much less arbitrary.

**Comment by ghost** [comment]
@cheeseonamonkey basically, when doing API requests you need to ensure that your total message array fits the context size in tokens. And you also want to control the length of the assistant response - `max_tokens`.

So, to have a really accurate calculation, you add messages to the context as long as the total token count of the context (`messages` array to send to the OAI API) is less than `max_context_size - max_tokens`, e.g. if `max_context_size = 4096` and `max_tokens = 300`, you should only add 3796 tokens' worth of context maximum.

The frontend already does prune the messages that don't fit the context, but it does that based on the "Max Tokens" option currently, which is misleading.

As far as I know the size of the most recent message doesn't really matter, usually in chats you'd have a lot of messages before your context gets filled.

**Comment by jackschedel** [comment]
@Yardanico rather than having it limit past messages to (max_context - max_tokens) like you suggested, I'm just going to have it be limited to max_context.
Max Context: How much past chat history to give
Max Tokens: How much to generate

**Comment by jackschedel** [comment]
**Max Tokens Working**
![image](https://github.com/jackschedel/gpt.koaladev.io/assets/26682594/fb3a5a87-98c8-4b93-ac69-f41375807ef4)


**Context Working**
![image](https://github.com/jackschedel/gpt.koaladev.io/assets/26682594/d69b914a-7101-4a4d-88b3-11e5cb036a2d)
(a few chats between)
![image](https://github.com/jackschedel/gpt.koaladev.io/assets/26682594/5736ff38-7997-4082-a722-e2779b194b6f)

**Comment by ghost** [comment]
thanks a lot!!

**Comment by ghost** [comment]
Sorry to annoy a bit, but there's a tiny bug with the current implementation - the actual max context should be the specified value by user minus max tokens for generation. Otherwise if you have e.g. a full context chat and have max_tokens set to 500, the response will likely be really short, because the model gets passed e.g 7900 tokens of context, so it can only generate 292 more - the model can't operate on more than its context .

Tldr - subtract `max_tokens` from the max context value when pruning context. 


I'll PR myself tomorrow probably, not at PC right now.

**Comment by jackschedel** [comment]
@Yardanico **have you confirmed that this is actually the case, or did you just look at my implementation?** I know that I subtracted it inside of the limit tokens function (rather than directly in the call to it like you did), as I ran into issues when doing it like you suggested. (but there's a chance I did it wrong obv)

I just tried an example with messages longer than the context length, max context set the model's max token context, max tokens set to very high, and asked it to write me 5 paragraphs, which it did; afaik it works.

**Comment by ghost** [comment]
@jackschedel sorry, my description was a bit wrong, but the issue is there. For example, with https://transfer.sh/ipEwfo8Io7/2023-08-14%20%281%29.json (a chat with 4008 tokens of context, 765 max tokens and 4096 max context with normal turbo) if you press "Generate", the request will be sent with `max_tokens` at 55, while `max_tokens` should always stay the same as in chat settings, the context should get pruned instead.

![image](https://github.com/jackschedel/KoalaClient/assets/21169548/063252a9-3ff3-4520-9df7-b43eb9e7f9f3)


The offending code is https://github.com/jackschedel/KoalaClient/blob/main/src/api/api.ts#L57

Thinking about it, I guess you could justify both approaches - lowering `max_tokens` or pruning more context to always have space for `max_tokens`, but from what I've seen most tools seem to do the latter, and it'll be more intuitive to the user - otherwise it could be hard to understand that `max_tokens` was lowered because of the big context.

Oh, and btw, have you thought of removing the fork status of the repository? That'll improve its discoverability and some other minor things. You can do it easily via support (it's actually automated) - https://support.github.com/request/fork

**Comment by jackschedel** [comment]
Working on a fix now.

Good idea on the fork detach, I didn't realize that it de-prioritized it in search results. Just submitted the request.

**Comment by jackschedel** [comment]
(@Yardanico )
